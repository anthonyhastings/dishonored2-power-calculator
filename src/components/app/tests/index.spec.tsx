import renderer from 'react-test-renderer';
import { useAppDispatch } from 'store-hooks';
import App from '../';

jest.mock('react-router-dom', () => ({
  Switch: 'MockSwitch',
  Route: 'MockRoute',
  Link: 'MockLink',
}));

jest.mock('selectors', () => ({
  hasInitialDataFailedSelector: jest.fn().mockReturnValue(false),
  isInitialDataIncompleteSelector: jest.fn().mockReturnValue(true),
}));

jest.mock('store-hooks', () => ({
  useAppSelector: (fn: () => void) => fn(),
  useAppDispatch: jest.fn(),
}));

jest.mock('slices/characters', () => ({
  fetchCharacters: () => 'MockFetchCharactersAction',
}));

jest.mock('slices/powers', () => ({
  fetchPowers: () => 'MockFetchPowersAction',
}));

jest.mock('components/loader', () => 'MockLoader');

jest.mock(
  'components/character-selection',
  () =>
    function MockCharacterSelection() {
      return <p>Mock Character Selection</p>;
    }
);

jest.mock(
  'components/page-not-found',
  () =>
    function MockPageNotFound() {
      return <p>Page Not Found</p>;
    }
);

jest.mock(
  'components/power-selection',
  () =>
    function MockPowerSelection() {
      return <p>Mock Power Selection</p>;
    }
);

describe('App component', () => {
  let testContext: {
    component?: renderer.ReactTestRenderer;
    dispatch: jest.Mock;
  };

  const mockedUseAppDispatch = useAppDispatch as jest.Mock;

  const renderComponent = () => <App />;

  beforeEach(() => {
    testContext = {
      dispatch: jest.fn(),
    };

    mockedUseAppDispatch.mockReturnValue(testContext.dispatch);
  });

  afterEach(() => {
    jest.clearAllMocks();
  });

  describe('when rendered', () => {
    beforeEach(async () => {
      await renderer.act(async () => {
        testContext.component = renderer.create(renderComponent());
      });
    });

    it('triggers calls for initial fetching of data', () => {
      expect(testContext.dispatch.mock.calls).toEqual([
        ['MockFetchCharactersAction'],
        ['MockFetchPowersAction'],
      ]);
    });

    it('renders a header area and top-level routes', () => {
      expect(testContext.component).toMatchSnapshot();
    });
  });
});
