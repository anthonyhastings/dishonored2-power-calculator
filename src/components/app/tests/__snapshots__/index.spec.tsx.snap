// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App component when rendered renders a header area and top-level routes 1`] = `
<div
  className="app"
>
  <header
    className="app__header"
  >
    <h1
      className="app__title"
    >
      Power Calculator
    </h1>
    <MockLink
      className="app__logo-link"
      to="/"
    >
      <img
        alt="Dishonored 2 logo"
        className="app__logo-graphic"
        draggable="false"
        height="90"
        sizes="100vw"
        srcSet="660x90.png 660w, 1320x180.png 1320w"
        width="660"
      />
    </MockLink>
  </header>
  <div
    className="app__content"
  >
    <MockLoader
      showError={false}
      showLoader={true}
    >
      <MockSwitch>
        <MockRoute
          exact={true}
          path="/"
        >
          <p>
            Mock Character Selection
          </p>
        </MockRoute>
        <MockRoute
          path="/powers/:characterSlug"
        >
          <p>
            Mock Power Selection
          </p>
        </MockRoute>
        <MockRoute>
          <p>
            Page Not Found
          </p>
        </MockRoute>
      </MockSwitch>
    </MockLoader>
  </div>
</div>
`;
